### PR TITLE
RFC Added SkipMissing wrapper metric

### PR DESCRIPTION
**Pull request by rofinn**
A potential solution to https://github.com/JuliaStats/Distances.jl/issues/11

Pros:

- Logic is isolated to a few dozen lines in 1 file
- Independent of distance metric types
- Seems like a pretty common approach elsewhere (ex [scikit learn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.nan_euclidean_distances.html))
- `skipmissing` already exists in base, so hopefully this approach is relatively intuitive

Cons:

- Performance costs (ie: generates a mask and views on each call)
- Sufficiently different from `Base.skipmissing` that we shouldn't export and could cause confusion?
- Arguably shouldn't be a metric and could probably just loosen other APIs to take a user defined function that calls `Missings.skipmissings`? (ex [NearestNeighbors.jl](https://github.com/KristofferC/NearestNeighbors.jl/blob/master/src/NearestNeighbors.jl#L26C81-L26C81))

### CODE REVIEW

File: src/missing.jl [hunk 1]
```diff
@@ -0,0 +1,47 @@
+"""
+Exclude any missing indices from being included the wrappped distance metric.
+"""
+struct SkipMissing{D<:PreMetric} <: PreMetric
```
Comment (rofinn) [inline src/missing.jl:4]: I'm not sure if there's a benefit to have specific subtypes here? All the metrics I tested seemed to work.

File: src/missing.jl [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    @inbounds for i in 1:n
+        mask[i] = !(ismissing(a[i]) || ismissing(b[i]))
+    end
+
+    # Calling `_evaluate` allows us to also mask metric parameters like weights or periods
+    # I don't think this can be generalized to user defined metric types though without knowing
+    # what the parameters mean.
+    # NTOE: We call disallowmissings to avoid downstream type promotion issues.
+    if dist.d isa UnionMetrics
+        params = parameters(dist.d)
```
Comment (rofinn) [inline src/missing.jl:30]: AFAICT, `parameters` is only defined for `UnionMetrics`?

File: src/missing.jl [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+        return _evaluate(
+            dist.d,
+            disallowmissing(view(a, mask)),
+            disallowmissing(view(b, mask)),
+            isnothing(params) ? params : view(params, mask),
+        )
+    else
+        return dist.d(
```
Comment (rofinn) [inline src/missing.jl:39]: I'm not sure if there's a safer fallback?

File: test/test_dists.jl [hunk 4]
```diff
@@ -1051,6 +1051,124 @@ end
...
+            # Doesn't handle eltype Union{T, Missing}
+            @test_throws MethodError colwise(dist, A[:, 3:4], B[:, 3:4])
+        else
+            @test ismissing(dist(a, b))
+            @test_throws MethodError colwise(dist, A, B)
+            @test_throws MethodError pairwise(dist, A, B, dims=2)
+            @test_throws MethodError pairwise(dist, A, dims=2)
+        end
+
+        # Handle weights
```
Comment (rofinn) [inline test/test_dists.jl:1154]: We're just testing that `skipmissing` returns the same results as deleting indices missing in either vector.

File: test/test_dists.jl [hunk 5]
```diff
@@ -1051,6 +1051,124 @@ end
...
+        if dist isa Distances.UnionMetrics && Distances.parameters(dist) isa Vector
+            @test D(a, b) == Distances._evaluate(dist, [1, 5], [6, 10], Distances.parameters(dist)[[1, 5]])
+        else
+            @test D(a, b) == dist([1, 5], [6, 10])
+        end
+
+        @test colwise(D, A, B)[3:4] ≈ colwise(dist, disallowmissing(A[:, 3:4]), disallowmissing(B[:, 3:4]))
```
Comment (rofinn) [inline test/test_dists.jl:1161]: Double check results are the same for non-missing columns.

File: test/test_dists.jl [hunk 6]
```diff
@@ -1051,6 +1051,124 @@ end
...
+
+        M = pairwise(D, A, B, dims=2)
+        @test size(M) == (4, 4)
+        @test !any(ismissing, M)
+
+        M = pairwise(D, A, dims=2)
+        @test size(M) == (4, 4)
+        @test !any(ismissing, M)
```
Comment (rofinn) [inline test/test_dists.jl:1169]: Could manually check the calculations here, but I figured a few smoke tests that it doesn't error or return `missing`s was good enough?

**Comment by rofinn** [comment]
Hmm, looks like some of the metrics throw different errors on different Julia versions.
